src/win/fs.c

```diff
@@ -1611,26 +1613,356 @@ static void fs__access(uv_fs_t* req) {
+                            DACL_SECURITY_INFORMATION;
+  if (ERROR_SUCCESS != GetNamedSecurityInfoW(req->file.pathw, SE_FILE_OBJECT, si, &psidOwner, &psidGroup, &pOldDACL, NULL, &pSD)) {
+    DWORD err = GetLastError();
+    SET_REQ_WIN32_ERROR(req, err);
+    goto chmod_cleanup;
+  }
+  
+  /* If the file does not contain a group owner, we will use the user's 'Creator Group ID' instead */
+  if (EqualSid(psidGroup, psidNull)) {
+    psidGroup = psidCreatorGroup;
+  }
+
+  /*
+   * We next need to scan all groups that the current user "belongs" to, in order to
+   * set the permissions for those groups to be the same as the "group" bit; so first
+   * we collect a list of group PSIDs:
+   */
+  DWORD numGroups = 0;
+  HANDLE hToken = NULL;
+  DWORD tokenAccess = TOKEN_IMPERSONATE |
+                      TOKEN_QUERY |
+                      TOKEN_DUPLICATE |
+                      STANDARD_RIGHTS_READ;
+  if (!OpenProcessToken(GetCurrentProcess(), tokenAccess, &hToken )) {
+    SET_REQ_WIN32_ERROR(req, GetLastError());
+    goto chmod_cleanup;
+  }
+  HANDLE hImpersonatedToken = NULL;
+  if (!DuplicateToken(hToken, SecurityImpersonation, &hImpersonatedToken)) {
+    SET_REQ_WIN32_ERROR(req, GetLastError());
+    goto chmod_cleanup;
+  }
+
+  /* Extract EAs from old DACL */
+  ULONG numOldEAs = 0;
+  ULONG numOtherGroups = 0;
+  if (ERROR_SUCCESS != GetExplicitEntriesFromAclW(pOldDACL, &numOldEAs, &poldEAs)) {
+    SET_REQ_WIN32_ERROR(req, GetLastError());
+    goto chmod_cleanup;
+  }
+
+  /* Iterate over all old ACEs, looking for groups that we belong to */
+  for (ULONG ea_idx=0; ea_idx<numOldEAs; ++ea_idx) {
+    PSID pEASid = (PSID)poldEAs[ea_idx].Trustee.ptstrName;
+    /* Skip this EA if it isn't an SID, or it is "Everyone" or our actual group */
+    if (poldEAs[ea_idx].Trustee.TrusteeForm != TRUSTEE_IS_SID ||
+        EqualSid(pEASid, psidEveryone) ||
+        EqualSid(pEASid, psidGroup)) {
+      continue;
+    }
+
+    /* Check to see if our user is a member of this group */
+    BOOL isMember = FALSE;
+    if (!CheckTokenMembership(hImpersonatedToken, pEASid, &isMember)) {
+      SET_REQ_WIN32_ERROR(req, GetLastError());
+      goto chmod_cleanup;
+    }
+
+    /* If we're a member, then count it */
+    if (isMember) {
+      numOtherGroups++;
+    }
+  }
+
+  /*
+   * Create an ACE for each triplet (user, group, other).
+   * Note that there is a priority mismatch between POSIX permissions and Windows ACLS:
+   * POSIX permissions are specificity-primary, in that a more specific permission takes
+   * precedence over a less specific one.  E.g. 0600 (allow user but deny group) will still
+   * end up allowing, whereas 0060 (deny user but allow group) will end up denying, because
+   * the "user" triplet is more specific than the group triplet.
+   * 
+   * In Windows ACLs, it is deny-primary; in that a deny anywhere in the ACL will result
+   * in a deny; therefore we build our masks by OR-ing in the more specific permissions with
```

Comment:
I think you're right; my previous experiments with were confused in the past because of the difference in behavior of `SetEntriesInAcl` with regards to allow/deny entries;

> The `SetEntriesInAcl` function places any new access-denied ACEs at the beginning of the list of ACEs for the new ACL. This function places any new access-allowed ACEs just before any existing access-allowed ACEs.

The problem with this (and why I thought it was `deny-primary` is that you can add an allow for an owner, but the pre-existing deny for a group will still override it, due to the group deny matching first, since the allows get added after the denies.  Example:

```
$ icacls foo
foo Everyone:(DENY)(S,X)
    NULL SID:(DENY)(Rc,S,REA,X,DC)
    WIN64BOT2\Administrator:(F)
    WIN64BOT2\None:(RX)
    NT AUTHORITY\SYSTEM:(RX,W,DC)
    BUILTIN\Administrators:(RX,W,DC)
    Everyone:(R)

$ ./julia/usr/bin/julia.exe -e '@show Sys.isexecutable("foo")'
Sys.isexecutable("foo") = false

$ icacls foo /grant "Everyone:(X)" /inheritance:r
processed file: foo
Successfully processed 1 files; Failed processing 0 files

$ icacls foo
foo NULL SID:(DENY)(Rc,S,REA,X,DC)
    WIN64BOT2\Administrator:(F)
    WIN64BOT2\None:(RX)
    NT AUTHORITY\SYSTEM:(RX,W,DC)
    BUILTIN\Administrators:(RX,W,DC)
    Everyone:(RX)

$ ./julia/usr/bin/julia.exe -e '@show Sys.isexecutable("foo")'
Sys.isexecutable("foo") = true
```

There still is no "specificity-primaryness" in the sense of owner always taking precedence over group; it's all position-dependent.  I'm not sure what to do here, since the "canonical" order is to put the denies first; in order for our "allow" to make an impact, it seems that we must eliminate the denies.